### DISCUSSION TITLE
Annotations for variable length function composition/pipeline.

### DISCUSSION THREAD

**Discussion by randolf-scholz** [python/typing#1766] [Q&A]
```python
class Transform(Protocol[X, Y]):
    def __call__(self, x: X, /) -> Y: ...

class Pipeline(Transform):
    funcs: list[Transform]  # <-- how to annotate?
    
    def __init__(self, funcs: Iterable[Transform], /) -> None:  # <-- how to annotate?
        self.funcs = list(funcs)

    def __call__(self, x):
        for func in self.funcs:
            x = func(x)
        return x
```

Is there any simple way to have more precise type hints for `Pipeline`, without going completely overboard with overloads?
In particular, it would be nice to have:

1. `Pipeline(Transform[X_first, Y_last])`
2. some type checking that the domain/codomain of each transform is compatible
3. Handling of the edge cases with 0 or 1 arguments: `Pipeline([])` and `Pipeline([Transform[X,Y]])`.

The best I could come up with handles (1) and (3), but requires changing the `__init__`-signature from `(self, funcs: Iterable[Transform], /)` to `self, *funcs: *tuple[Transform, ...], /)`, and the introduction of an additional subclass `IdentityPipe`.

I wonder if there is any simpler/better way of doing it.

-----

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYA0UAKogKYBqAhiDWAG4MgVgsATGgEEUcGqRYBnGDRANuLCgH14CBjQByCnjWlSeMVYwBQJgBpQAvHUasQACgBE5pwEoTATWu319557uJgCqPvR%2BbM7BQUxhdpFOTEEmpBT6dCAsKFLAuBAORGAk5BQA2uY0ngC6bgBcJlCNUAIMwFDKypIUKsoOBhTANAAetVAVUAD0blAAtAB8UJ6jAHSrZqnpBEjqFKgMDrSZ2bkgEKXBNEw19U1QAAJcPHyCDU0tbR0oDADuHQ6pUhoACpgABXFCkKSjIEwUEICgMUoONw1WYLJwASRaKBgSHgW3UThWa1uD24vH4AlejXe7WUX1%2BvQBwLBEKhUBhcIRpUOWRyeXKlSqqPmUCcBIYuy%2BgsWVSJUFWy2p90eFJet1pnx%2Bf2ZHNZkOhsPhiN5xwF41EcCqwKN3NN-NONEV1oyfJOZ0tQpF6IlUsR42q8sVys19O1TIogL14INHNtJqODogTtW10aAGIoAg4CAkGgABYwUZ5lC4RHyHB4DEQsDZJAyBg4gDy5OeAiqytu8lhIBQUCkcJ4yOWWsZ-0jLJjUg8ypDrTpqDxf36g2jbMNXITbryKeW3qgmlrDFGUEzAB4ZjMoPmwN9YGAoFlSzAWDAGAB%2BTtNFfLfVSHy7DIDh-jOGrzh0XQ9H0kqriMUAXJM0yikwNy3E0Qw%2BJIQGelAQweGhTQnFA%2BqYH2P5-qhBG3BhNj6g4eFfo03agr2UBYTADhMMMM4bFI-5Yo2uL4ts%2By%2BnspQ4Za1zKhRUCATApTaOSHZgR8nRKFBK7DKM5hIQs5iUWhzGsUMZgAMqjPa7qlKgsjEc8MBVABLAQAARgILC4aMwAOfRHgAFqWYm1k%2Bfwdmha%2BTk2GkbkeV5UAAExAqZtBBdupylBFdkyCAUVyS57meXBOV%2BWYcG2T4ACMZgEAADD4YlfMiJj6IYxjqAUtU0AJOJ4nAEoeK1eDtaJtV%2BTQtkzgQlUNSJfoOGZg18W1aiiZVNCNYitk0FlNQtctw2rQUlXjfZYVTQls07HsC00LQS0GIdjAFAlG1zeJ239jAuUPStz0EAlp05VNADMV2SjdZk0P5d2-U9HUECDb3XdKn05XtQ1GEdiNA99HhAA)

```python
from typing import Protocol, TypeVar, overload, Any, cast, reveal_type, Never, assert_type

X = TypeVar("X")
Y = TypeVar("Y")
U = TypeVar("U")
V = TypeVar("V")

class Transform(Protocol[X, Y]):
    def __call__(self, x: X, /) -> Y: ...

class Pipeline(Transform[U, V]):
    @overload
    def __new__(cls, *funcs: *tuple[()]) -> "IdentityPipe": ...
    @overload
    def __new__(cls, *funcs: *tuple[Transform[X, Y]]) -> "Pipeline[X, Y]": ...
    @overload
    def __new__(cls, *funcs: *tuple[Transform[X, Any], *tuple[Transform, ...], Transform[Any, Y]]) -> "Pipeline[X, Y]": ...
    def __new__(cls, *funcs: *tuple[Transform, ...]):  # pyright: ignore[reportInconsistentOverload]
        return super().__new__(cls, *funcs)

    
    def __init__(self, *funcs: *tuple[Transform, ...]) -> None:  # <-- how to annotate?
        self.funcs = list(funcs)

    def __call__(self, x: U, /) -> V:
        x = cast(Any, x)
        for func in self.funcs:
            x = func(x)
        return cast(V, x)

class IdentityPipe(Pipeline[Any, Any]):
    funcs: list[Never]
    def __call__(self, x: X) -> X:
        return x

S: Transform[int, float] = lambda x: float(x)
Z: Transform[float, float] = lambda x: 2*x
T: Transform[float, str] = lambda x: str(x)

x: int = 1

P0 = Pipeline()
assert_type(P0, IdentityPipe)
assert_type(P0(x), int)

P1 = Pipeline(S)
assert_type(P1, Pipeline[int, float])
assert_type(P1(x), float)

P2 = Pipeline(S, T)
assert_type(P2, Pipeline[int, str])
assert_type(P2(x), str)

P3 = Pipeline(S, Z, T)
assert_type(P3, Pipeline[int, str])
assert_type(P3(x), str)

```

**Comment by decorator-factory** [comment]
Right now that is not possible. However, you can chain method calls to achieve the same result:
```py
from typing import Generic, TypeVar
from collections.abc import Callable

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")

class Pipe(Generic[A, B]):
    def __init__(self, fn: Callable[[A], B]) -> None:
        self._fn = fn

    def __call__(self, arg: A, /) -> B:
        return self._fn(arg)
    
    def then(self, fn: Callable[[B], C], /) -> "Pipe[A, C]":
        return Pipe(lambda a: fn(self._fn(a)))
```
Then you can do `pipe = Pipe(S).then(Z).then(T)` and have `pipe` inferred as `Pipe[int, str]`

**Reply by randolf-scholz** [reply]
I do something like this, in my code the `Transform`-derived base class has a method for chaining that returns `Pipeline(self, other)`
